src/ArrayComponents.ml

```diff
@@ -0,0 +1,51 @@
+open Base
+
+open Expr
+
+let all = [
+  {
+    name = "select";
+    codomain = Type.TVAR ("'b");
+    domain = [Type.ARRAY (Type.TVAR "'a", Type.TVAR "'b"); Type.TVAR "'a"];
+    is_argument_valid = (function
+                         | _ -> true);
+    evaluate = (function [@warning "-8"]
+                [Value.Array (_, _, elems, default_val) ; key]
+                -> match List.Assoc.find elems ~equal:Value.equal key with
+                   | None -> default_val
+                   | Some value -> value);
+    to_string = (fun [@warning "-8"] [a ; b] -> "(select " ^ a ^ " " ^ b ^ ")");
+    global_constraints = (fun _ -> [])
+  }
+  ;
+  {
+    name = "store";
+    codomain = Type.ARRAY (Type.TVAR "'a", Type.TVAR "'b");
+    domain = [Type.ARRAY (Type.TVAR "'a", Type.TVAR "'b"); Type.TVAR "'a"; Type.TVAR "'b"];
+    is_argument_valid = (function
+                         | _ -> true);
+    evaluate = (function [@warning "-8"]
+                [Value.Array (key_type, val_type, elems, default_val) ; key; value]
+                -> Value.Array (key_type,val_type, (key, value)::elems,default_val));
+    to_string = (fun [@warning "-8"] [a ; b ; c] -> "(store " ^ a ^ " " ^ b ^ " " ^ c ^ ")");
+    global_constraints = (fun _ -> [])
+  }
+
+  ;
+  {
+    name = "equal";
+    codomain = Type.BOOL;
+    domain = [Type.ARRAY (Type.TVAR "'a", Type.TVAR "'b"); Type.ARRAY (Type.TVAR "'a", Type.TVAR "'b")];
+    is_argument_valid = (function
+                         | _ -> true);
+    evaluate = (function [@warning "-8"]
+                [Value.Array (a_key_type, a_val_type, a_elems, a_default_val) ; Value.Array (b_key_type, b_val_type, b_elems, b_default_val)]
+                -> Value.Bool (    (Poly.equal a_key_type b_key_type)
+                                && (Poly.equal a_val_type b_val_type)
+                                && (Poly.equal a_default_val b_default_val)
+                                && (Poly.equal a_elems b_elems)));
```

Comment:
This isn't correct -- specifically the `Poly.equal a_elems b_elems` part. The following would return false as per this implementation:

```
(equal (store (store arr k1 v1) k2 v2)
       (store (store arr k2 v2) k1 v1))
```

Reply:
The _ideal_ solution is to change `Value.Array` to store a `HashTable` instead of `List` -- then, the order of stores doesn't matter (and the lookup during `select` would also be *much* faster: amortized O(1) vs O(n)).

I would suggest this:
1. Leave a `FIXME` at `Value.Array` definition, for example: `(* FIXME: Use HashTable instead of List *)`
2. To fix this `equal` implementation, use `Poly.equal` after [sorting](https://ocaml.janestreet.com/ocaml-core/latest/doc/core_kernel/Core_kernel/List/index.html#val-sort) both arrays

Reply:
Ok, yea the poly.equal will fail in your example. Hashtable makes more sense, I will leave a fixme. 
However, equality component will still be broken after sorting and using Poly.equal. 
Consider the following example:
Let's say we have two arrays `a` and `b`. We do the following;  `(= a (store (store a 0 1) 0 2))` , and `(= b (store b 0 2))`. These two arrays are equal and z3 would return `sat`.  However, the OCaml implementation of arrays appends the `(key, value)` to the top of the list. So we will have `a = [(0,2) , (0,1)]` and b would be `[(0,2)]`. This will fail the sort and poly.equal check?

Reply:
Good catch! Use [`dedup_and_sort`](https://ocaml.janestreet.com/ocaml-core/latest/doc/core_kernel/Core_kernel/List/index.html#val-dedup_and_sort) instead.

Note that, even building a HashTable just for this function isn't trivial. You have to more sure that any key type is hashable, and that needs more work. Leave a fixme for now and let's not introduce HashTables in this PR.